### ISSUE TITLE
Random number generation (random_data) query

### ISSUE THREAD

**Issue by michaellrowley** [itzmeanjan/kyber#23]
I've been working to implement this library (great work by the way, I don't see many projects with such great documentation and consistency without large amounts of users) into a personal project but after reading your [``random_data`` function](https://github.com/itzmeanjan/kyber/blob/bfd9bbbbea6ee064c92bec212cef6058db5c1139/include/utils.hpp#L12-L25) based in ``utils.hpp`` I was left unsure about its suitability for generating large quantities of data outside of its use as a generator for KYBER.
From what I can gather, you're seeding a MT19937 instance with the output of a single ``random_device`` generation and then generate the random data using MT19937.
My concern is that this would mean that if the function was called with a long enough ``len`` argument and an attacker could access a substantial section of the returned buffer, they could recreate the internal state of the MT19937 instance and determine the rest of the data generated by that instance (thus, the rest of that call to ``random_data``).

I'm not 100% sure if that is correct and if it is; this still isn't an issue with your usage as the mersenne instance is reseeded for every call and the buffers filled by those calls are never 'partially' revealed to an adversary (i.e via a public key) such that the rest being generated would be an issue. However, in my instance, if I needed to generate a large buffer (large enough to be susceptible to an internal-state recreation attack) and then reveal a chunk of it to an assumed attacker, would I be better off using another solution such as the underlying ``random_device``, or OpenSSL's ``RAND_bytes()``?

TL;DR: Would you recommend that [this](https://github.com/itzmeanjan/kyber/blob/master/include/utils.hpp#L12-L25) is used as a CSPRNG for large quantities of data outside of this library.

**Comment by itzmeanjan** [comment]
Hi @michaellrowley ,

> My concern is that this would mean that if the function was called with a long enough `len` argument and an attacker could access a substantial section of the returned buffer, they could recreate the internal state of the MT19937 instance and determine the rest of the data generated by that instance (thus, the rest of that call to `random_data`).

I'm not very sure about this, but I won't discard the possibility too.

---

I think it's better if you use  OpenSSL's `Rand_bytes`, just to be on the sure side. 

But I've another suggestion. You can use `std::random_device` to seed MT19937 engine and then collect 32 -bytes random seed from it, which you can now pass to SHAKE256 hasher and squeeze out how ever many bytes you may need for filling your large buffer. SHAKE256 being an XoF, seems to be helpful in this case.

If you're looking for one SHAKE256 implementation, I've got one https://github.com/itzmeanjan/sha3 .
To be specific you may want to look at https://github.com/itzmeanjan/sha3#usage .
